File: packages/pptx/src/proposals.ts
import type {
  DeckSnapshot,
  ParagraphAlignment,
  ShapeRect,
  ShapeSnapshot,
  ShapeStroke,
  SlideDisplayList,
  TextStyleSnapshot,
} from './types';

export type ProposalEdit =
  | {
      type: 'replaceText';
      storyId: string;
      start: number;
      end: number;
      text: string;
      style?: Partial<TextStyleSnapshot> | null;
    }
  | {
      type: 'formatText';
      storyId: string;
      start: number;
      end: number;
      patch: Partial<TextStyleSnapshot>;
    }
  | {
      type: 'setParagraphAlignment';
      storyId: string;
      start: number;
      end: number;
      alignment: ParagraphAlignment | 'justLow' | 'dist' | 'thaiDist' | null;
    }
  | { type: 'setShapeRect'; slideId: string; shapeId: string; rect: ShapeRect }
  | {
      type: 'setShapeFill';
      slideId: string;
      shapeId: string;
      color: string | null;
    }
  | {
      type: 'setShapeStroke';
      slideId: string;
      shapeId: string;
      stroke: { [K in keyof ShapeStroke]: ShapeStroke[K] | null };
    }
  | {
      type: 'setShapeAdjust';
      slideId: string;
      shapeId: string;
      adjustments: Record<string, number>;
    }
  | { type: 'setSlideNotes'; slideId: string; text: string };

export interface ProposalChange {
  slideId: string;
  shapeId: string | null;
  before: ShapeSnapshot | null;
  after: ShapeSnapshot | null;
  oldText: string;
  newText: string;
}

export interface Proposal {
  id: string;
  agentId: string;
  note: string | null;
  edits: ProposalEdit[];
  changes: ProposalChange[];
  staleTargets: string[];
}

export interface ProposalPreview {
  proposal: Proposal;
  snapshot: DeckSnapshot;
}

export interface ProposalTextChange {
  storyId: string;
  start: number;
  end: number;
  kind: 'insertion' | 'deletion';
}

export interface ProposalDiffSlide extends ProposalPreview {
  frame: SlideDisplayList;
  textChanges: ProposalTextChange[];
}

export interface ProposalAcceptance {
  proposalId: string;
  applied: boolean;
  snapshot: DeckSnapshot;
}

export class StaleProposalError extends Error {
  readonly targets: string[];

  constructor(targets: string[]) {
    super(`Proposal targets changed: ${targets.join(', ')}`);
    this.name = 'StaleProposalError';
    this.targets = targets;
  }
}

File: bindings/python-pptx/python/betteroffice_pptx/proposals.py
"""Pending edits and rendered-preview metadata."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class ProposalChange:
    slide_id: str
    shape_id: str | None
    before: dict[str, Any] | None
    after: dict[str, Any] | None
    old_text: str
    new_text: str

    @classmethod
    def _from_dict(cls, value: dict[str, Any]) -> ProposalChange:
        return cls(
            value["slideId"], value["shapeId"], value["before"], value["after"],
            value["oldText"], value["newText"],
        )


@dataclass(frozen=True)
class Proposal:
    id: str
    agent_id: str
    note: str | None
    edits: tuple[dict[str, Any], ...]
    changes: tuple[ProposalChange, ...]
    stale_targets: tuple[str, ...]

    @classmethod
    def _from_dict(cls, value: dict[str, Any]) -> Proposal:
        return cls(
            value["id"], value["agentId"], value["note"], tuple(value["edits"]),
            tuple(ProposalChange._from_dict(change) for change in value["changes"]),
            tuple(value["staleTargets"]),
        )


@dataclass(frozen=True)
class ProposalPreview:
    proposal: Proposal
    snapshot: dict[str, Any]

    @classmethod
    def _from_dict(cls, value: dict[str, Any]) -> ProposalPreview:
        return cls(Proposal._from_dict(value["proposal"]), value["snapshot"])
